src/Pos/Block/Network/Retrieval.hs

```diff
@@ -109,19 +110,29 @@ retrievalWorkerImpl sendActions =
                                else pure Nothing
             case (mbQueuedHeadersChunk, mbRecHeader) of
                 (Nothing, Nothing) -> retry
-                (Just chunk, _)    -> pure (handleBlockRetrieval chunk)
-                (_, Just rec')     -> pure (handleHeadersRecovery rec')
+                (Just (nodeId, task), _) ->
+                    case task of
+                        RetrieveBlocksByHeaders headers ->
+                            pure (handleBlockRetrieval nodeId headers)
+                        RetrieveHeadersByTip tip ->
+                            pure $ do
+                                mghM <- mkHeadersRequest (Just (headerHash tip))
+                                whenJust mghM $ \mgh ->
+                                    withConnectionTo sendActions nodeId $ \_ -> pure $ Conversation $
+                                        requestHeaders mgh nodeId (Just tip)
```

Comment:
How I understand this logic, requestHeaders will add another message to queue to download blocks. Which is wrong, we should download block right away